[numpy] `torch.sin` : support and promote integer inputs to float

kshitij12345 opened the pull request:
References #42515

> Enable integer -> float unary type promotion for ops like sin

Will follow-up for other such Ops once this PR is merged.

cc: @mruberry

kshitij12345 committed:
add iterator support for promoting unary ops to float

kshitij12345 committed:
update test

kshitij12345 commented:
Are there other places to update tests at?

mruberry commented:
> Are there other places to update tests at?

I think a new property for the unary ufuncs, like "promotes_integers_to_float", and new tests in test_unary_ufuncs.py for the type promotion behavior would be helpful. These tests could be:

- integer inputs -> float outputs
- inplace on integer tensors does not work
- out with integer inputs does not work, out with float inputs does work (and the out tensor values are equal to the output of the function)

kshitij12345 committed:
fix impl

kshitij12345 committed:
update test support

* assert inplace on integer tensors do not work
  for unary ops which promote integer to float.
* verify against numpy for integer input.

kshitij12345 committed:
update property name

kshitij12345 committed:
fix stray arg name

kshitij12345 committed:
call iter.cast_outputs()

kshitij12345 committed:
add test to test out with different dtypes

kshitij12345 commented on test/test_unary_ufuncs.py:
```diff
@@ -204,11 +204,20 @@ def _fn(t):
+                with self.assertRaises(RuntimeError):
+                    alt(t.clone())
+                continue
+
```
> inplace on integer tensors does not work

kshitij12345 commented on test/test_unary_ufuncs.py:
```diff
@@ -270,7 +279,17 @@ def test_reference_numerics(self, device, dtype, op):
+                                           exact_dtype=exact_dtype, rtol=1e-4, atol=1e-3)
+                    continue
+
+            self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)
```
> integer inputs -> float outputs

kshitij12345 commented on test/test_unary_ufuncs.py:
```diff
@@ -377,6 +396,61 @@ def test_batch_vs_slicing(self, device, dtype, op):
+                op(input, out=output)
+
+    @ops(unary_ufuncs)
+    def test_out_arg_all_dtypes(self, device, dtype, op):
```
> out with integer inputs does not work, out with float inputs does work (and the out tensor values are equal to the output of the function)

kshitij12345 reviewed (commented):
Have updated the test to verify the above mentioned 3 points and marked the relevant update in tests as comments. 
PTAL :)

kshitij12345 committed:
update support for torch.bool

kshitij12345 committed:
fix function name

kshitij12345 committed:
fix correct variable name

kshitij12345 committed:
update out_dtype to include bool and half

kshitij12345 committed:
Merge branch 'master' into develop/numpy/unary_float_ops

kshitij12345 commented:
Gentle Ping :)

kshitij12345 committed:
skip nan_to_num for out arg test

nan_to_num currently has incorrect behaviour
and breaks out arg tets.

kshitij12345 committed:
fix nan_to_num behavior and update test to verify it

kshitij12345 commented:
@mruberry PTAL:)

kshitij12345 commented:
@mruberry Gentle Ping:)

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -315,8 +323,13 @@ Tensor& sgn_out(Tensor& result, const Tensor& self) {
-Tensor sin(const Tensor& self) { return unary_op_impl(self, at::sin_out); }
+Tensor& sin_out(Tensor& result, const Tensor& self) { return unary_op_impl_float_out(result, self, sin_stub); }
+Tensor sin(const Tensor& self) {
+  Tensor result;
```
What about another helper, like unary_op_impl_float?

mruberry commented on test/test_unary_ufuncs.py:
```diff
@@ -205,11 +205,20 @@ def _fn(t):
         expected = op(t)
 
-        for alt in (op.get_method(), op.get_inplace(), torch.jit.script(_fn)):
+        for inplace, alt in ((False, op.get_method()), (True, op.get_inplace()),
```
This is good but reverse the order, "alt, inplace"

mruberry commented on test/test_unary_ufuncs.py:
```diff
@@ -220,7 +229,7 @@ def assertEqualHelper(self, actual, expected, msg, *, dtype, exact_dtype=True, *
         # Some NumPy functions return scalars, not arrays
         if isinstance(expected, Number):
-            self.assertEqual(actual.item(), expected)
+            self.assertEqual(actual.item(), expected, **kwargs)
```
Nice fix

mruberry commented on test/test_unary_ufuncs.py:
```diff
@@ -271,7 +280,17 @@ def test_reference_numerics(self, device, dtype, op):
+            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
+                exact_dtype = False
+
+                if dtype in [torch.uint8, torch.int8, torch.bool]:
```
Can you elaborate on what's going on here?

kshitij12345 replied:
Without this we get an error due to tolerance.
```
============================================================================== FAILURES ==============================================================================
______________________________________________________ TestUnaryUfuncsCPU.test_reference_numerics_sin_cpu_bool _______________________________________________________

self = <test_unary_ufuncs.TestUnaryUfuncsCPU testMethod=test_reference_numerics_sin_cpu_bool>, device = 'cpu', dtype = torch.bool
op = <torch.testing._internal.common_methods_invocations.UnaryUfuncInfo object at 0x7f130193d7f0>

    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
    @suppress_warnings
    @ops(unary_ufuncs)
    def test_reference_numerics(self, device, dtype, op):
        include_extremals = (op.handles_complex_extremals if
                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
    
        tensors = generate_numeric_tensors(device, dtype,
                                           domain=op.domain,
                                           include_large_values=op.handles_large_floats,
                                           include_extremal_values=include_extremals)
        for t in tensors:
            if dtype is torch.bfloat16:
                a = t.cpu().to(torch.float32).numpy()
            else:
                a = t.cpu().numpy()
    
            actual = op(t)
            expected = op.ref(a)
    
            # Crafts a custom error message for smaller, printable tensors
            if t.numel() < 10:
                msg = ("Failed to produce expected results! Input tensor was"
                       " {0}, torch result is {1}, and reference result is"
                       " {2}.").format(t, actual, expected)
            else:
                msg = None
    
            exact_dtype = True
            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
                exact_dtype = False
    
                # if dtype in [torch.uint8, torch.int8, torch.bool]:
                    # Facing tolerance issue with above dtypes.
                    # self.assertEqualHelper(actual, expected, msg, dtype=dtype,
                    #                        exact_dtype=exact_dtype, rtol=1e-4, atol=1e-3)
                    # continue
    
>           self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)

test/test_unary_ufuncs.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_unary_ufuncs.py:232: in assertEqualHelper
    self.assertEqual(actual.item(), expected, **kwargs)
../PytorchENV/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py:1174: in assertEqual
    self.assertTrue(result, msg=msg)
E   AssertionError: False is not true : Scalars failed to compare as equal! Comparing 0.8414709568023682 and 0.84130859375 gives a difference of 0.00016236305236816406, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only 1.1093701171875e-05!
______________________________________________________ TestUnaryUfuncsCPU.test_reference_numerics_sin_cpu_int8 _______________________________________________________

self = <test_unary_ufuncs.TestUnaryUfuncsCPU testMethod=test_reference_numerics_sin_cpu_int8>, device = 'cpu', dtype = torch.int8
op = <torch.testing._internal.common_methods_invocations.UnaryUfuncInfo object at 0x7f130193d7f0>

    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
    @suppress_warnings
    @ops(unary_ufuncs)
    def test_reference_numerics(self, device, dtype, op):
        include_extremals = (op.handles_complex_extremals if
                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
    
        tensors = generate_numeric_tensors(device, dtype,
                                           domain=op.domain,
                                           include_large_values=op.handles_large_floats,
                                           include_extremal_values=include_extremals)
        for t in tensors:
            if dtype is torch.bfloat16:
                a = t.cpu().to(torch.float32).numpy()
            else:
                a = t.cpu().numpy()
    
            actual = op(t)
            expected = op.ref(a)
    
            # Crafts a custom error message for smaller, printable tensors
            if t.numel() < 10:
                msg = ("Failed to produce expected results! Input tensor was"
                       " {0}, torch result is {1}, and reference result is"
                       " {2}.").format(t, actual, expected)
            else:
                msg = None
    
            exact_dtype = True
            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
                exact_dtype = False
    
                # if dtype in [torch.uint8, torch.int8, torch.bool]:
                    # Facing tolerance issue with above dtypes.
                    # self.assertEqualHelper(actual, expected, msg, dtype=dtype,
                    #                        exact_dtype=exact_dtype, rtol=1e-4, atol=1e-3)
                    # continue
    
>           self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)

test/test_unary_ufuncs.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_unary_ufuncs.py:232: in assertEqualHelper
    self.assertEqual(actual.item(), expected, **kwargs)
../PytorchENV/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py:1174: in assertEqual
    self.assertTrue(result, msg=msg)
E   AssertionError: False is not true : Scalars failed to compare as equal! Comparing -0.8414709568023682 and -0.84130859375 gives a difference of 0.00016236305236816406, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only 1.1093701171875e-05!
______________________________________________________ TestUnaryUfuncsCPU.test_reference_numerics_sin_cpu_uint8 ______________________________________________________

self = <test_unary_ufuncs.TestUnaryUfuncsCPU testMethod=test_reference_numerics_sin_cpu_uint8>, device = 'cpu', dtype = torch.uint8
op = <torch.testing._internal.common_methods_invocations.UnaryUfuncInfo object at 0x7f130193d7f0>

    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
    @suppress_warnings
    @ops(unary_ufuncs)
    def test_reference_numerics(self, device, dtype, op):
        include_extremals = (op.handles_complex_extremals if
                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
    
        tensors = generate_numeric_tensors(device, dtype,
                                           domain=op.domain,
                                           include_large_values=op.handles_large_floats,
                                           include_extremal_values=include_extremals)
        for t in tensors:
            if dtype is torch.bfloat16:
                a = t.cpu().to(torch.float32).numpy()
            else:
                a = t.cpu().numpy()
    
            actual = op(t)
            expected = op.ref(a)
    
            # Crafts a custom error message for smaller, printable tensors
            if t.numel() < 10:
                msg = ("Failed to produce expected results! Input tensor was"
                       " {0}, torch result is {1}, and reference result is"
                       " {2}.").format(t, actual, expected)
            else:
                msg = None
    
            exact_dtype = True
            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
                exact_dtype = False
    
                # if dtype in [torch.uint8, torch.int8, torch.bool]:
                    # Facing tolerance issue with above dtypes.
                    # self.assertEqualHelper(actual, expected, msg, dtype=dtype,
                    #                        exact_dtype=exact_dtype, rtol=1e-4, atol=1e-3)
                    # continue
    
>           self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)

test/test_unary_ufuncs.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_unary_ufuncs.py:232: in assertEqualHelper
    self.assertEqual(actual.item(), expected, **kwargs)
../PytorchENV/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py:1174: in assertEqual
    self.assertTrue(result, msg=msg)
E   AssertionError: False is not true : Scalars failed to compare as equal! Comparing 0.8414709568023682 and 0.84130859375 gives a difference of 0.00016236305236816406, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only 1.1093701171875e-05!
_____________________________________________________ TestUnaryUfuncsCUDA.test_reference_numerics_sin_cuda_bool ______________________________________________________

self = <test_unary_ufuncs.TestUnaryUfuncsCUDA testMethod=test_reference_numerics_sin_cuda_bool>, device = 'cuda:0', dtype = torch.bool
op = <torch.testing._internal.common_methods_invocations.UnaryUfuncInfo object at 0x7f130193d7f0>

    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
    @suppress_warnings
    @ops(unary_ufuncs)
    def test_reference_numerics(self, device, dtype, op):
        include_extremals = (op.handles_complex_extremals if
                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
    
        tensors = generate_numeric_tensors(device, dtype,
                                           domain=op.domain,
                                           include_large_values=op.handles_large_floats,
                                           include_extremal_values=include_extremals)
        for t in tensors:
            if dtype is torch.bfloat16:
                a = t.cpu().to(torch.float32).numpy()
            else:
                a = t.cpu().numpy()
    
            actual = op(t)
            expected = op.ref(a)
    
            # Crafts a custom error message for smaller, printable tensors
            if t.numel() < 10:
                msg = ("Failed to produce expected results! Input tensor was"
                       " {0}, torch result is {1}, and reference result is"
                       " {2}.").format(t, actual, expected)
            else:
                msg = None
    
            exact_dtype = True
            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
                exact_dtype = False
    
                # if dtype in [torch.uint8, torch.int8, torch.bool]:
                    # Facing tolerance issue with above dtypes.
                    # self.assertEqualHelper(actual, expected, msg, dtype=dtype,
                    #                        exact_dtype=exact_dtype, rtol=1e-4, atol=1e-3)
                    # continue
    
>           self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)

test/test_unary_ufuncs.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_unary_ufuncs.py:232: in assertEqualHelper
    self.assertEqual(actual.item(), expected, **kwargs)
../PytorchENV/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py:1174: in assertEqual
    self.assertTrue(result, msg=msg)
E   AssertionError: False is not true : Scalars failed to compare as equal! Comparing 0.8414710164070129 and 0.84130859375 gives a difference of 0.00016242265701293945, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only 1.1093701171875e-05!
_____________________________________________________ TestUnaryUfuncsCUDA.test_reference_numerics_sin_cuda_int8 ______________________________________________________

self = <test_unary_ufuncs.TestUnaryUfuncsCUDA testMethod=test_reference_numerics_sin_cuda_int8>, device = 'cuda:0', dtype = torch.int8
op = <torch.testing._internal.common_methods_invocations.UnaryUfuncInfo object at 0x7f130193d7f0>

    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
    @suppress_warnings
    @ops(unary_ufuncs)
    def test_reference_numerics(self, device, dtype, op):
        include_extremals = (op.handles_complex_extremals if
                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
    
        tensors = generate_numeric_tensors(device, dtype,
                                           domain=op.domain,
                                           include_large_values=op.handles_large_floats,
                                           include_extremal_values=include_extremals)
        for t in tensors:
            if dtype is torch.bfloat16:
                a = t.cpu().to(torch.float32).numpy()
            else:
                a = t.cpu().numpy()
    
            actual = op(t)
            expected = op.ref(a)
    
            # Crafts a custom error message for smaller, printable tensors
            if t.numel() < 10:
                msg = ("Failed to produce expected results! Input tensor was"
                       " {0}, torch result is {1}, and reference result is"
                       " {2}.").format(t, actual, expected)
            else:
                msg = None
    
            exact_dtype = True
            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
                exact_dtype = False
    
                # if dtype in [torch.uint8, torch.int8, torch.bool]:
                    # Facing tolerance issue with above dtypes.
                    # self.assertEqualHelper(actual, expected, msg, dtype=dtype,
                    #                        exact_dtype=exact_dtype, rtol=1e-4, atol=1e-3)
                    # continue
    
>           self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)

test/test_unary_ufuncs.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_unary_ufuncs.py:232: in assertEqualHelper
    self.assertEqual(actual.item(), expected, **kwargs)
../PytorchENV/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py:1174: in assertEqual
    self.assertTrue(result, msg=msg)
E   AssertionError: False is not true : Scalars failed to compare as equal! Comparing -0.8414710164070129 and -0.84130859375 gives a difference of 0.00016242265701293945, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only 1.1093701171875e-05!
_____________________________________________________ TestUnaryUfuncsCUDA.test_reference_numerics_sin_cuda_uint8 _____________________________________________________

self = <test_unary_ufuncs.TestUnaryUfuncsCUDA testMethod=test_reference_numerics_sin_cuda_uint8>, device = 'cuda:0', dtype = torch.uint8
op = <torch.testing._internal.common_methods_invocations.UnaryUfuncInfo object at 0x7f130193d7f0>

    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
    @suppress_warnings
    @ops(unary_ufuncs)
    def test_reference_numerics(self, device, dtype, op):
        include_extremals = (op.handles_complex_extremals if
                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
    
        tensors = generate_numeric_tensors(device, dtype,
                                           domain=op.domain,
                                           include_large_values=op.handles_large_floats,
                                           include_extremal_values=include_extremals)
        for t in tensors:
            if dtype is torch.bfloat16:
                a = t.cpu().to(torch.float32).numpy()
            else:
                a = t.cpu().numpy()
    
            actual = op(t)
            expected = op.ref(a)
    
            # Crafts a custom error message for smaller, printable tensors
            if t.numel() < 10:
                msg = ("Failed to produce expected results! Input tensor was"
                       " {0}, torch result is {1}, and reference result is"
                       " {2}.").format(t, actual, expected)
            else:
                msg = None
    
            exact_dtype = True
            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
                exact_dtype = False
    
                # if dtype in [torch.uint8, torch.int8, torch.bool]:
                    # Facing tolerance issue with above dtypes.
                    # self.assertEqualHelper(actual, expected, msg, dtype=dtype,
                    #                        exact_dtype=exact_dtype, rtol=1e-4, atol=1e-3)
                    # continue
    
>           self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)

test/test_unary_ufuncs.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_unary_ufuncs.py:232: in assertEqualHelper
    self.assertEqual(actual.item(), expected, **kwargs)
../PytorchENV/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py:1174: in assertEqual
    self.assertTrue(result, msg=msg)
E   AssertionError: False is not true : Scalars failed to compare as equal! Comparing 0.8414710164070129 and 0.84130859375 gives a difference of 0.00016242265701293945, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only 1.1093701171875e-05!
====================================================================== short test summary info =======================================================================
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_reference_numerics_sin_cpu_bool - AssertionError: False is not true : Scalars failed to compare as equal...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_reference_numerics_sin_cpu_int8 - AssertionError: False is not true : Scalars failed to compare as equal...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_reference_numerics_sin_cpu_uint8 - AssertionError: False is not true : Scalars failed to compare as equa...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_sin_cuda_bool - AssertionError: False is not true : Scalars failed to compare as equ...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_sin_cuda_int8 - AssertionError: False is not true : Scalars failed to compare as equ...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_sin_cuda_uint8 - AssertionError: False is not true : Scalars failed to compare as eq...
========================================================== 6 failed, 236 passed, 2052 deselected in 15.99s ===========================================================
```

kshitij12345 replied:
On further investigating,
Pytorch defaults output to `float` for the following dtypes while numpy to `float16`
```python
>>> a = torch.randn(3,3) > 0.5
>>> a
tensor([[ True, False,  True],
        [ True, False,  True],
        [ True, False, False]])
>>> a.float().sin()
tensor([[0.8415, 0.0000, 0.8415],
        [0.8415, 0.0000, 0.8415],
        [0.8415, 0.0000, 0.0000]])
>>> import numpy as np
>>> np.sin(a.numpy())
array([[0.8413, 0.    , 0.8413],
       [0.8413, 0.    , 0.8413],
       [0.8413, 0.    , 0.    ]], dtype=float16) # Note float16
```

kshitij12345 replied:
Sure.

kshitij12345 replied:
Done.

mruberry commented on test/test_unary_ufuncs.py:
```diff
@@ -378,6 +397,61 @@ def test_batch_vs_slicing(self, device, dtype, op):
+        if self.device_type == 'cuda':
+            # As input are dynamically casted
+            # in kernel, cast input to out_dtype
+            expected = op(input.to(out_dtype))
```
This is not generally true. For example, op(fp16, out=fp64) != op(fp64). The former computation occurs in float16, while the latter computation occurs in the wider, higher precision float64.

kshitij12345 replied:
Explanation below.

mruberry commented on test/test_unary_ufuncs.py:
```diff
@@ -378,6 +397,61 @@ def test_batch_vs_slicing(self, device, dtype, op):
+            expected = op(input.to(out_dtype))
+            op(input, out=out)
+            self.assertEqual(out, expected)
+        else:
```
There should be no observable different based on the device type.

kshitij12345 replied:
I think the difference is because when it is CUDA, the input is dynamically cast to output type correctly.
So if input is `int64` and output is `double`, then `int64` is converted to `double` and output is saved in `double` tensor.

However on CPU, `int64` is converted to `float` as that is the computed common_dtype when integer is promoted and then we explicitly call `iter.cast_outputs()` (as discussed as per the offline conversation) to cast it to `double`.

Reference : https://github.com/pytorch/pytorch/blob/a277c097acc615b4710d615387891b7f764df326/aten/src/ATen/native/TensorIterator.cpp#L345-L347

I was facing tolerance issue on tests otherwise.

Let me know what you think.

mruberry replied:
The common dtype is computed without considering the output, however, and the computation should occur in the common dtype. @zasdfgbnm, is there a bug in our CUDA tensor iterator loops where they're computing in the output dtype?

mruberry replied:
Let's look at some example programs.

```
a = 7.3333484203230823
b = 2.44040844

print(a/b)

# Creates a double scalar tensor representable in half precision
a_t = torch.tensor(a, dtype=torch.double).half().double()
b_t = torch.tensor(b, dtype=torch.double).half().double()

print(a_t / b_t)

out = torch.empty(0, dtype=torch.double)
torch.div(a_t.half(), b_t.half(), out=out)
print(out)
```

So this program divides two values in Python, double, and then in half but with a double output. It prints:

```
3.0049676521865667
tensor(3.0056, dtype=torch.float64)
tensor(3.0059, dtype=torch.float64)
```

This is expected. Note the divergence between the double and half inputs. Also note that this divergence is not a representational issue with inputs because of the half->double bobble performed earlier.

Now let's do this on CUDA:

```
a = 7.3333484203230823
b = 2.44040844

print(a/b)

a_t = torch.tensor(a, dtype=torch.double).half().double().cuda()
b_t = torch.tensor(b, dtype=torch.double).half().double().cuda()

print(a_t / b_t)

out = torch.empty(0, dtype=torch.double).cuda()
torch.div(a_t.half(), b_t.half(), out=out)
print(out)
```

The program is the same except the tensors are placed on a CUDA device. This prints:

```
3.0049676521865667
tensor(3.0056, device='cuda:0', dtype=torch.float64)
tensor(3.0059, device='cuda:0', dtype=torch.float64)
```

The same values. So it appears that the output dtype is correctly not affecting the computation on CPU or CUDA for div.

zasdfgbnm replied:
IIUC TensorIterator GPU loops allow doing computing in any dtype. The dynamic casting mechanism will handle the cast automatically. You can even do
float --(cast)--> int --(compute)--> int --(cast)--> float, 
nothing will prevent you from doing so, and there will be no error message. It is the responsibility of the operator implementor to ensure the correct type is used.
From https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/UnaryGeometricKernels.cu, we can see that lots of computations are done in output type, instead of common type.

zasdfgbnm replied:
If we agree that computation should be done in common type instead of output type, then this file should be fixed.

mruberry replied:
Agreed. I've only fixed the behavior in TensorIterator, and I'm sure other parts of our system still use the output argument to determine what computation they do. Analogously, our proposed updates to out resizing behavior were implemented in TensorIterator but few other places. Would you file an issue where you notice the out type affecting the computation, @zasdfgbnm?

zasdfgbnm replied:
https://github.com/pytorch/pytorch/issues/46302

mruberry replied:
> #46302

Thank you!

mruberry replied:
OK, that conversation aside, what is going on in this particular case that the CPU and CUDA code paths are divergent, @kshitij12345? Is there an error with setting the appropriate computation dtype? Or a numerical precision issue on CUDA?

kshitij12345 replied:
It is the case of numerical difference.

mruberry replied:
Oh, in that case would you just set the precision high enough for both of them?

kshitij12345 replied:
Is it fine if we update the CUDA computation type to common_dtype type as mentioned in the conversation above. 
I have verified locally that it solves the issue.
```C++
void sin_kernel_cuda(TensorIterator& iter) {
  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(ScalarType::Half, iter.common_dtype(), "sin_cuda", [&]() {
    gpu_kernel(iter, []GPU_LAMBDA(scalar_t a) -> scalar_t {
      return ::sin(a);
    });
  });
}
```

kshitij12345 replied:
@mruberry Do let me know if this is fine.

mruberry replied:
Yes! I was just about to come here and say exactly that. We should fix this as @zasdfgbnm suggests above.

kshitij12345 replied:
Great. Done!

mruberry commented on test/test_unary_ufuncs.py:
```diff
@@ -378,6 +397,61 @@ def test_batch_vs_slicing(self, device, dtype, op):
+            out = torch.empty_like(input, dtype=out_dtype)
+            if op.promotes_integers_to_float:
+                self._test_out_promote_int_to_float_op(op, input, out)
+                continue
```
else: instead of continue here

mruberry commented on test/test_unary_ufuncs.py:
```diff
@@ -378,6 +397,61 @@ def test_batch_vs_slicing(self, device, dtype, op):
+        dtype = input.dtype
+        out_dtype = output.dtype
+        if out_dtype.is_floating_point and not dtype.is_complex:
+            self._compare_out_variant_based_on_device(op, input, output)
```
Is there a test that sin(int, out=fp64) works? That is, the float32 output should be able to be cast to (weakly) wider types, like double and cfloat.

mruberry commented:
> @mruberry Gentle Ping:)

Thanks for the ping ;).

kshitij12345 committed:
add unary_op_impl_float helper

kshitij12345 committed:
address commit

* change order alt, inplace.
* use else.

kshitij12345 committed:
Merge branch 'master' into develop/numpy/unary_float_ops

kshitij12345 commented:
@mruberry 
Error on TE Fuser is real.

I don't have much knowledge on Tensor Expression code. Would be great if I get some pointers
```
======================================================================
ERROR [0.931s]: test_unary_ops (__main__.TestTEFuser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_jit_fuser_te.py", line 1307, in test_unary_ops
    torch.testing.assert_allclose(ref, t(x))
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 127, in prof_func_call
    return prof_callable(func_call, *args, **kwargs)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 124, in prof_callable
    return callable(*args, **kwargs)
RuntimeError: default_program(22): error: more than one instance of overloaded function "sin" matches the argument list:
            function "sin(double)"
            function "sin(float)"
            argument types are: (char)

1 error detected in the compilation of "default_program".

nvrtc compilation failed: 

#define NAN __int_as_float(0x7fffffff)
#define POS_INFINITY __int_as_float(0x7f800000)
#define NEG_INFINITY __int_as_float(0xff800000)


template<typename T>
__device__ T maximum(T a, T b) {
  return isnan(a) ? a : (a > b ? a : b);
}

template<typename T>
__device__ T minimum(T a, T b) {
  return isnan(a) ? a : (a < b ? a : b);
}

extern "C" __global__
void func_172(char* t0, double* aten_sin_flat) {
{
  if (512 * blockIdx.x + threadIdx.x<2 ? 1 : 0) {
    int v = (int)(__ldg(t0 + 512 * blockIdx.x + threadIdx.x));
    aten_sin_flat[512 * blockIdx.x + threadIdx.x] = (double)(sin((char)(2 * v)));
  }
}
}


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_jit_fuser_te.py", line 1311, in test_unary_ops
    " ".join(["Failed:", str(dtype), op.__name__, device])
RuntimeError: Failed: torch.int8 sin cuda

----------------------------------------------------------------------
Ran 62 tests in 34.828s

FAILED (errors=1, skipped=14)
```

ngimel commented:
cc @ZolotukhinM for TE failure. I guess we should just disable the test.

ZolotukhinM commented:
cc @nickgg who recently fixed some dtype issues in TE cuda codegen

ZolotukhinM commented:
> I guess we should just disable the test.
It's dangerous in this case. Do I understand correctly that this PR changes the behavior of `torch.sin`? If that's the case, the TE lowering part should be updated accordingly. What happens here is that TE lowers it to an expression without int8->something promotion and compiler errors out.

Most probably, this code needs to be updated: https://github.com/pytorch/pytorch/blob/e7dbaa252e85779660c1c13262340ae2b891c7b4/torch/csrc/jit/tensorexpr/kernel.cpp#L946-L949

And we will now need to generate a `Cast` expression there for the `sin` argument.

ngimel commented:
This is a wider problem of how TE reproduces (or does not) type promotion behavior of eager mode. Are there any documents to this effect?
If there is not a general strategy, I'd advise against point fixes for separate functions, and bail out of fusing ops where type promotion is required. If type promotion is actually supported and follows eager rules, great!

ZolotukhinM commented:
Our overall strategy so far has been to match exactly what eager mode does and treat any divergence as a bug - the test in question has caught exactly that. AFAIK, there is no document describing it, but we should have tests checking that (I think we verify product of all supported ops and all supported dtypes, but I might be mistaken here - @bertmaher or @nickgg could probably tell for sure).

mruberry commented:
> Our overall strategy so far has been to match exactly what eager mode does and treat any divergence as a bug - the test in question has caught exactly that. AFAIK, there is no document describing it, but we should have tests checking that (I think we verify product of all supported ops and all supported dtypes, but I might be mistaken here - @bertmaher or @nickgg could probably tell for sure).

What's the plan when eager changes its behavior, though? Disabling the corresponding TE tests? ccing someone? Filing an issue?

ZolotukhinM commented:
> What's the plan when eager changes its behavior, though? Disabling the corresponding TE tests? ccing someone? Filing an issue?

I'd advice against disabling TE tests since TE is now enabled in JIT by default and disabling without fixing would result in JIT diverging from eager. Whenever eager changes its behavior we need to update the TE  correspondingly since it's modeling that behavior. Btw, there are actually two ways to approach this from the TE side: we can either stop fusing the modified op, or we can adjust how it's modeled in the TE. I've mostly been talking about the latter option, but the former might often be easier.

How often does it usually happen? If not very often, asking someone from the TE/NNC to catch up with the new behavior sounds good to me (i.e. ccing + filing an issue). If that happens often, we probably need to come up with a more efficient approach.

mruberry commented:
> I'd advice against disabling TE tests since TE is now enabled in JIT by default and disabling without fixing would result in JIT diverging from eager. 

Right.

> Whenever eager changes its behavior we need to update the TE correspondingly since it's modeling that behavior. Btw, there are actually two ways to approach this from the TE side: we can either stop fusing the modified op, or we can adjust how it's modeled in the TE. I've mostly been talking about the latter option, but the former might often be easier.

The former does sound easier. A real fix for this would be for there to be a common shape prop or type promotion system that would dictate operations' eager mode and NNC behavior. cc @ezyang 

> How often does it usually happen? If not very often, asking someone from the TE/NNC to catch up with the new behavior sounds good to me (i.e. ccing + filing an issue). If that happens often, we probably need to come up with a more efficient approach.

Semi-often? There's a good chance we enable int->float type promotion for most unary ufuncs this release.

ZolotukhinM commented:
> There's a good chance we enable int->float type promotion for most unary ufuncs this release.

We're talking about 1.8 here, right? Let's try to do the CC+file an issue approach for now and revisit it if it gets too messy.

mruberry commented:
> > There's a good chance we enable int->float type promotion for most unary ufuncs this release.
> 
> We're talking about 1.8 here, right? Let's try to do the CC+file an issue approach for now and revisit it if it gets too messy.

Yes, in 1.8. Sure.

ezyang commented:
Type promotion semantics are documented at https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc This doc, however, doesn't seem to be updated with promotion semantics for integer inputs to unary floating point operations, which is what is in question here. Promotion in TE is implemented at https://github.com/pytorch/pytorch/blob/e7dbaa252e85779660c1c13262340ae2b891c7b4/torch/csrc/jit/tensorexpr/kernel.cpp#L384 (unrelatedly, it doesn't seem like TE is handling promotion of scalar inputs differently)

It seems to me that TE just needs to know which unary ops promote integer inputs to float, and just insert necessary promotion checks everywhere. This doesn't seem like rocket science.

kshitij12345 committed:
update commit related for precision relaxation

kshitij12345 committed:
sin cuda to use common_dtype and update relevant test

mruberry commented on test/test_unary_ufuncs.py:
```diff
@@ -271,7 +280,18 @@ def test_reference_numerics(self, device, dtype, op):
+
+                if dtype in [torch.uint8, torch.int8, torch.bool]:
+                    # For the above dtypes,
+                    # `torch` prmotes and computes output in `float` while numpy in `float16`
```
"prmotes" -> "promotes"

kshitij12345 committed:
address comment

Rephrase the note for numpy-pytorch type promotion
difference for uint8, int8, bool

mruberry commented:
@kshitij12345 I think this looks really good, thanks for another awesome PR. 

@ZolotukhinM What do you want to do to land this PR? It could comment out this line and cite an issue:

https://github.com/pytorch/pytorch/blob/ec5f81f9d334efa3d46ca9c66885e2ab14b6ce01/test/test_jit_fuser_te.py#L1277

ZolotukhinM commented:
> @ZolotukhinM What do you want to do to land this PR? It could comment out this line and cite an issue:

Yes, something along these lines would do. That would probably break some tests in `test_tensorexpr.py` and/or `test_jit_fuser_te.py` - please feel free to temporarily disable them. We will re-enable them once the promotion is properly handled.

mruberry commented:
I filed https://github.com/pytorch/pytorch/issues/46458. @kshitij12345, would you comment out that one line in the TE tests and add a comment to see this issue?

kshitij12345 committed:
Merge branch 'master' into develop/numpy/unary_float_ops

kshitij12345 committed:
disable sin in te fuser test

kshitij12345 committed:
Merge branch 'develop/numpy/unary_float_ops' into develop/numpy/unary_float_ops

kshitij12345 commented:
@mruberry Thanks for filing the issue . Have commented and added reference in test. PTAL:)

kshitij12345 commented:
@mruberry Gentle Ping :)

mruberry reviewed (approved):
Nice work, @kshitij12345!

Now let's start supporting this behavior with other ops, too. I think these functions could also use an update:

torch.angle
torch.exp
torch.log
torch.log2
torch.log10
torch.expm1
torch.log1p
torch.sqrt
torch.reciprocal (this would be divergent from NumPy, but that's OK)
torch.sin
torch.cos
torch.tan
torch.asin
torch.acos
torch.atan
torch.sinh
torch.cosh
torch.tanh
torch.erf
torch.erfc
torch.erfinv
torch.rsqrt
torch.sigmoid
torch.digamma
torch.lgamma
torch.polygamma

There are probably a few more, too. 
torch.mvlgamma

kshitij12345 commented:
@mruberry Sure. Thanks for the list!
